pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java

```diff
@@ -744,6 +745,21 @@ public void close() throws PulsarClientException {
         }
     }
 
+    private void closeUrlLookupMap() {
+        Map<String, LookupService> closedUrlLookupServices = new HashMap(urlLookupMap.size());
+        urlLookupMap.entrySet().forEach(e -> {
+            try {
+                e.getValue().close();
+            } catch (Exception ex) {
+                log.error("Error closing lookup service {}", e.getKey(), ex);
+            }
+            closedUrlLookupServices.put(e.getKey(), e.getValue());
+        });
+        closedUrlLookupServices.entrySet().forEach(e -> {
```

Comment:
Why did you retain the lookup services that failed to `close`? It seems that there is no chance to close these lookup services again.

I mean, what's the issue with the following simple implementation.

```java
    private void closeUrlLookupMap() {
        urlLookupMap.forEach((serviceUrl, lookupService) -> {
            try {
                lookupService.close();
            } catch (Exception e) {
                log.error("Error closing lookup service {}", serviceUrl, e);
            }
        });
        urlLookupMap.clear();
    }
```

Reply:
yes. agree

Reply:
> Why did you retain the lookup services that failed to close? It seems that there is no chance to close these lookup services again.

Just left the information, so we can do a troubleshooting through a client-side heap dump 😂 
PS: troubleshooting is always difficult through logs. The more information there, the easier troubleshooting is

Reply:
It makes sense. But it's better to add a comment to explain it. Otherwise other contributors might just clear this map in future.

```java
    private void closeUrlLookupMap() {
        final Map<String, LookupService> failedLookupMap = urlLookupMap.entrySet().stream().filter(__ -> {
            try {
                __.getValue().close();
                return false;
            } catch (Exception e) {
                log.error("Error closing lookup service {}", __.getKey(), e);
                return true;
            }
        }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
        urlLookupMap.clear();
        urlLookupMap.putAll(failedLookupMap); // retain the failed lookup services in memory for deeper debugging
    }
```

How about this implementation above?

Reply:
```java
1. urlLookupMap.clear();
2. urlLookupMap.putAll(failedLookupMap);
```

If we just look at the method `closeUrlLookupMap `, after `line-1` and before `line-2`, a command `map.computIfAbsent(k, v)` that using the same key in the map `failedLookupMap` may have been called, so it is not perfect. The current implementation is not simple to read, but it guarantees the correct logic.